PsfLauncher/main.cpp

```diff
@@ -268,43 +275,247 @@ void LaunchMonitorInBackground(std::filesystem::path packageRoot, const wchar_t
+            applicationNameForError = execInfo.CommandLine.substr(0, execInfo.CommandLine.find(' '));
+            ss << L"ERROR: Failed to create a process for " << applicationNameForError << " ";
+        }
+
+        ErrorInformation error = { ss.str(), err };
+        return error;
+    }
 }
-int __stdcall wWinMain(HINSTANCE, HINSTANCE, PWSTR args, int cmdShow)
+
+ErrorInformation StartWithShellExecute(std::filesystem::path packageRoot, std::filesystem::path exeName, std::wstring exeArgString, LPCWSTR dirStr, int cmdShow) noexcept
 {
-    return launcher_main(args, cmdShow);
+    // Non Exe case, use shell launching to pick up local FTA
+    auto nonexePath = packageRoot / exeName;
+
+    SHELLEXECUTEINFO shex = {
+        sizeof(shex)
+        , SEE_MASK_NOCLOSEPROCESS
+        , (HWND)nullptr
+        , nullptr
+        , nonexePath.c_str()
+        , exeArgString.c_str()
+        , dirStr ? (packageRoot / dirStr).c_str() : nullptr
+        , static_cast<WORD>(cmdShow)
+    };
+
+
+    Log("\tUsing Shell launch: %ls %ls", shex.lpFile, shex.lpParameters);
+    if (!ShellExecuteEx(&shex))
+    {
+        auto err{ ::GetLastError() };
+        std::wstring errorString = L"ERROR: Failed to create detoured shell process";
+
+        ErrorInformation error = { errorString, err };
+        return error;
+    }
+
+    ErrorInformation noError;
+    return noError;
 }
+
+static inline bool check_suffix_if(iwstring_view str, iwstring_view suffix) noexcept
+{
+    if ((str.length() >= suffix.length()) && (str.substr(str.length() - suffix.length()) == suffix))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+void Log(const char fmt[], ...) noexcept
+{
+    std::string str;
+    str.resize(256);
+
+    va_list args;
+    va_start(args, fmt);
+    std::size_t count = std::vsnprintf(str.data(), str.size() + 1, fmt, args);
+    assert(count >= 0);
+    va_end(args);
+
+    if (count > str.size())
+    {
+        str.resize(count);
+
+        va_list args2;
+        va_start(args2, fmt);
+        count = std::vsnprintf(str.data(), str.size() + 1, fmt, args2);
+        assert(count >= 0);
+        va_end(args2);
+    }
+
+    str.resize(count);
+    ::OutputDebugStringA(str.c_str());
+}
+void LogString(const char name[], const char value[]) noexcept
+{
+    Log("\t%s=%s\n", name, value);
+}
+void LogString(const char name[], const wchar_t value[]) noexcept
+{
+    Log("\t%s=%ls\n", name, value);
+}
+
+ErrorInformation CheckIfPowershellIsInstalled(bool& isPowershellInstalled) noexcept
+{
+    UniqueHKEY registryKey;
+    DWORD statusOfRegistryKey;
+    LSTATUS createResult = RegCreateKeyExW(HKEY_LOCAL_MACHINE, L"SOFTWARE\\Microsoft\\PowerShell\\1", 0, nullptr, REG_OPTION_NON_VOLATILE, KEY_READ, nullptr, &registryKey, &statusOfRegistryKey);
+
+    if (createResult != ERROR_SUCCESS)
+    {
+        ErrorInformation error = { L"Error with getting the key to see if PowerShell is installed. ", (DWORD)createResult };

```

Comment:
>ErrorInformation error =  [](start = 8, length = 25)

just return { ... }

Reply:
Is this for styling, or for performance?

Reply:
I did fix it by the way.